rust/src/dcerpc/dcerpc.rs

```diff
@@ -0,0 +1,2600 @@
+    transmute(boxed)
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_dcerpc_state_free(state: *mut std::os::raw::c_void) {
+    let _drop: Box<DCERPCState> = transmute(state);
+}
+
+#[no_mangle]
+pub extern "C" fn rs_dcerpc_state_transaction_free(_state: *mut std::os::raw::c_void, _tx_id: u64) {
+    // do nothing
+}
+
+#[no_mangle]
+pub extern "C" fn rs_dcerpc_get_tx_detect_state(
+    vtx: *mut std::os::raw::c_void,
+) -> *mut core::DetectEngineState {
+    let dce_state = cast_pointer!(vtx, DCERPCState);
+    match dce_state.de_state {
+        Some(ds) => ds,
+        None => std::ptr::null_mut(),
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn rs_dcerpc_set_tx_detect_state(
+    vtx: *mut std::os::raw::c_void,
+    de_state: *mut core::DetectEngineState,
+) -> u8 {
+    let dce_state = cast_pointer!(vtx, DCERPCState);
+    dce_state.de_state = Some(de_state);
+    0
+}
+
+#[no_mangle]
+pub extern "C" fn rs_dcerpc_get_tx(
+    state: *mut std::os::raw::c_void,
+    _tx_id: u64,
+) -> *mut DCERPCState {
+    let dce_state = cast_pointer!(state, DCERPCState);
+    dce_state
+}
+
+#[no_mangle]
+pub extern "C" fn rs_dcerpc_get_tx_cnt(_state: *mut std::os::raw::c_void) -> u8 {
+    1
+}
+
+#[no_mangle]
+pub extern "C" fn rs_dcerpc_get_alstate_progress(
+    _tx: *mut std::os::raw::c_void,
+    _direction: u8,
+) -> u8 {
+    0
+}
+
+#[no_mangle]
+pub extern "C" fn rs_dcerpc_get_alstate_progress_completion_status(_direction: u8) -> u8 {
+    1
+}
+
+#[no_mangle]
+pub extern "C" fn rs_dcerpc_get_tx_detect_flags(vtx: *mut std::os::raw::c_void, dir: u8) -> u64 {
+    let state = cast_pointer!(vtx, DCERPCState);
+    if dir & core::STREAM_TOSERVER != 0 {
+        return state.detect_flags_ts;
+    }
+    state.detect_flags_tc
+}
+
+#[no_mangle]
+pub extern "C" fn rs_dcerpc_set_tx_detect_flags(
+    vtx: *mut std::os::raw::c_void,
+    dir: u8,
+    flags: u64,
+) {
+    let state = cast_pointer!(vtx, DCERPCState);
+    if dir & core::STREAM_TOSERVER != 0 {
+        state.detect_flags_ts = flags;
+    } else {
+        state.detect_flags_tc = flags;
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_dcerpc_set_buffer(
+    state: &mut DCERPCState,
+    buf: *mut *const u8,
+    len: *mut u32,
+    dir: u8,
+) -> u8 {
+    match dir {
+        core::STREAM_TOSERVER => {
+            if let Some(ref req) = state.request {
+                *len = req.stub_data_buffer_len as u32;
+                *buf = req.stub_data_buffer.as_ptr();
+                SCLogDebug!("DCERPC Request stub buffer: Setting buffer to: {:?}", *buf);
+            }
+        }
+        _ => {
+            if let Some(ref resp) = state.response {
+                *len = resp.stub_data_buffer_len as u32;
+                *buf = resp.stub_data_buffer.as_ptr();
+                SCLogDebug!("DCERPC Response stub buffer: Setting buffer to: {:?}", *buf);
+            }
+        }
+    }
+    // TODO good thing to do? Returning endianness from the function whihch is not meant to do that
```

Comment:
maybe better to return through a function arg. Then let this function return whether it was able to provide a buffer to the caller at all.